sqlglot/optimizer/qualify_columns.py

```diff
@@ -0,0 +1,321 @@
+    """
+    expression = expression.copy()
+    _qualify_union(expression, schema, Sequence(), {})
+    return expression
+
+
+class Sequence:
+    """
+    Auto-incrementing sequence.
+
+    Example:
+        >>> seq = Sequence()
+        >>> seq.next()
+        0
+        >>> seq.next()
+        1
+
+    This is useful for autogenerating unique alias names.
+    """
+
+    def __init__(self):
+        self.i = -1
+
+    def next(self):
+        self.i += 1
+        return self.i
+
+
+def _qualify_union(expression, schema, sequence, parent_selectables):
+    """
+    Search UNION for columns to qualify.
+
+    Args:
+        expression (exp.Select or exp.Union): expression to search
+        schema (dict): Mapping of table names to all available columns.
+        sequence (Sequence): Auto-incrementing sequence.
+        parent_selectables (dict): Mapping of name to selectable columns.
+            This can include names set by CTEs or names that are available
+            from the parent context (e.g. in the case of correlated subqueries).
+    Returns:
+        list: output column names
+    """
+    if isinstance(expression, exp.Select):
+        return _qualify_select(expression, schema, sequence, parent_selectables)
+    if isinstance(expression, exp.Union):
+        left = _qualify_select(expression.this, schema, sequence, parent_selectables)
+        right = _qualify_union(
+            expression.args.get("expression"), schema, sequence, parent_selectables
+        )
+        if set(left) != set(right):
+            raise RuntimeError("UNION columns not equal")
+        return left
+
+    raise RuntimeError("Unexpected statement type")
+
+
+def _qualify_select(expression, schema, sequence, parent_selectables):
+    """
+    Search SELECT for columns to qualify.
+
+    Args:
+        (Same as `_qualify_union`)
+    Returns:
+        (Same as `_qualify_union`)
+    """
+    # pylint: disable=too-many-locals
+    ctes = []
+    derived_tables = []
+    subqueries = []
+    tables = []
+    columns = []
+    selections = []
+    select_stars = []
+
+    # Collect all the selections in this context
+    for selection in expression.args.get("expressions", []):
+        if isinstance(selection, exp.Star) or (
+            isinstance(selection, exp.Column) and isinstance(selection.this, exp.Star)
+        ):
+            select_stars.append(selection)
+        else:
+            selections.append(selection)
+
+    # Collect all the other relevant nodes in this context
+    # Only traverse down to the next SELECT statements - we'll recurse into those later
+    for node, parent, _ in expression.walk(stop_types=exp.Select):
+        if node is expression:
+            continue  # Skip this node itself - we only care about children
+        if isinstance(node, exp.CTE):
+            ctes.append(node)
+        elif isinstance(node, (exp.Select, exp.Union)):
+            if isinstance(parent, exp.Subquery):
```

Comment:
how come these are not stored as subqueries? curious about why these are called derived_tables

Reply:
I haven't quite figured out the naming here. I was borrowing from MySQL documentation.

```sql
SELECT * FROM x WHERE a IN (SELECT ...) <-- subquery
SELECT * FROM (SELECT ...)  <-- derived table... a special case of subqueries
```

Derived table add "selectables" to the current context.
The inverse happens with the other type of subquery... context get's pushed down, which support correlated subqueries.

I agree, the naming is super confusing here. I'll try to figure something out (and add some comments regardless).